nixos/modules/services/web-apps/cloudlog.nix

```diff
@@ -0,0 +1,395 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
+
+let
+  cfg = config.services.cloudlog;
+  dbFile = let
+    password = if cfg.database.createLocally
+               then "''"
+               else (if cfg.database.passwordFile != null
+                     then "trim(file_get_contents('${cfg.database.passwordFile}'))"
+                     else "'${cfg.database.password}'");
+  in pkgs.writeText "database.php" ''
+    <?php
+    defined('BASEPATH') OR exit('No direct script access allowed');
+    $active_group = 'default';
+    $query_builder = TRUE;
+    $db['default'] = array(
+      'dsn' => "",
+      'hostname' => '${cfg.database.host}',
+      'username' => '${cfg.database.user}',
+      'password' => ${password},
+      'database' => '${cfg.database.name}',
+      'dbdriver' => 'mysqli',
+      'dbprefix' => "",
+      'pconnect' => TRUE,
+      'db_debug' => (ENVIRONMENT !== 'production'),
+      'cache_on' => FALSE,
+      'cachedir' => "",
+      'char_set' => 'utf8mb4',
+      'dbcollat' => 'utf8mb4_general_ci',
+      'swap_pre' => "",
+      'encrypt' => FALSE,
+      'compress' => FALSE,
+      'stricton' => FALSE,
+      'failover' => array(),
+      'save_queries' => TRUE
+    );
+  '';
+  configFile = let
+    listToString = value:
+      "[ " + (concatStringsSep ", " (map valueToString value)) + "]";
+    valueToString = value: with builtins;
+      if typeOf value == "string" then "'${value}'"
+      else (if typeOf value == "bool" then (if value then "true"
+                                            else "false")
+            else (if typeOf value == "list" then listToString value
+                  else toString value));
+    cfgOptions = concatStringsSep "\n"
+      (map (n: let
+        value = cfg.settings."${n}";
+      in "$config['${n}'] = ${valueToString value};")
+        (attrNames cfg.settings));
```

Comment:
Are they using a custom config format?

Reply:
It is a php file. I've simplified this so that we use specific options to set the minimum required configuration, and added an extraConfig option that allows users to enter php config directly.